chainer/functions/connection/dropconnect.py

```diff
@@ -0,0 +1,100 @@
+from chainer import function
+from chainer.utils import type_check
+
+
+def _as_mat(x):
+    if x.ndim == 2:
+        return x
+    return x.reshape(len(x), -1)
+
+
+class DropConnect(function.Function):
+
+    """Linear unit using dropconnect regularization."""
+
+    def __init__(self, dropconnect_ratio):
+        self.dropconnect_ratio = dropconnect_ratio
+
+    def check_type_forward(self, in_types):
+        n_in = in_types.size()
+        type_check.expect(2 <= n_in, n_in <= 3)
+        x_type, w_type = in_types[:2]
+
+        type_check.expect(
+            x_type.dtype.kind == 'f',
+            w_type.dtype.kind == 'f',
+            x_type.ndim >= 2,
+            w_type.ndim == 2,
+            type_check.prod(x_type.shape[1:]) == w_type.shape[1],
+        )
+        if n_in.eval() == 3:
+            b_type = in_types[2]
+            type_check.expect(
+                b_type.dtype == x_type.dtype,
+                b_type.ndim == 1,
+                b_type.shape[0] == w_type.shape[0],
+            )
+
+    def forward(self, inputs):
+        scale = inputs[1].dtype.type(1. / (1 - self.dropconnect_ratio))
+        xp = cuda.get_array_module(*inputs)
+        if xp == numpy:
+            flag = xp.random.rand(*inputs[1].shape) >= self.dropconnect_ratio
+        else:
+            flag = (xp.random.rand(*inputs[1].shape, dtype=numpy.float32) >=
+                    self.dropconnect_ratio)
+        self.mask = scale * flag
+
+        x = _as_mat(inputs[0])
+        W = inputs[1] * self.mask
+        y = x.dot(W.T).astype(x.dtype, copy=False)
+        if len(inputs) == 3:
+            b = inputs[2]
+            y += b
+        return y,
+
+    def backward(self, inputs, grad_outputs):
+        x = _as_mat(inputs[0])
+        W = inputs[1] * self.mask
+        gy = grad_outputs[0]
+
+        gx = gy.dot(W).astype(x.dtype, copy=False).reshape(inputs[0].shape)
+        gW = gy.T.dot(x).astype(W.dtype, copy=False)
+        if len(inputs) == 3:
+            gb = gy.sum(0)
+            return gx, gW, gb
+        else:
+            return gx, gW
+
+
+def dropconnect(x, W, b=None, ratio=.5, train=True):
+    """Linear function using dropconnect.
```

Comment:
To me, "linear function *with* dropconnect" seems more appropriate because "using dropconnect" sounds like implementation of Linear function made from dropconnect.